nvflare/app_common/streamers/object_retriever.py

```diff
@@ -0,0 +1,284 @@
+        pass
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.START_RUN:
+            engine = fl_ctx.get_engine()
+            self.register_stream_processing(
+                fl_ctx=fl_ctx,
+                channel=self.stream_channel,
+                topic=self.topic,
+                stream_done_cb=self._handle_stream_done,
+            )
+            engine.register_aux_message_handler(topic=self.topic, message_handle_func=self._handle_request)
+
+    def retrieve(self, from_site: str, fl_ctx: FLContext, timeout: float, **obj_attrs) -> (str, Any):
+        """Retrieve an object from a specified site.
+
+        Args:
+            from_site: the site to retrieve the object from
+            fl_ctx: a FLContext object
+            timeout: max number of seconds to wait for the data
+            **obj_attrs: attributes of the object to be retrieved
+
+        Returns: tuple of (ReturnCode, Retrieved Data)
+
+        """
+        engine = fl_ctx.get_engine()
+        waiter = _Waiter()
+        tx_id = str(uuid.uuid4())
+        self.tx_table[tx_id] = waiter
+        self.log_debug(fl_ctx, f"set waiter for Rtr {tx_id}")
+
+        request = Shareable({RETRIEVER_TX_ID: tx_id})
+        if obj_attrs:
+            request.update(obj_attrs)
+
+        # ask the site to start streaming
+        reply = engine.send_aux_request(
+            targets=[from_site], request=request, topic=self.topic, fl_ctx=fl_ctx, timeout=timeout
+        )
+        # the reply is a dict keyed with site names!
+        reply = reply.get(from_site)
+
+        # now the reply is a Shareable object
+        if not isinstance(reply, Shareable):
+            self.log_error(fl_ctx, f"bad reply from site {from_site}: expect Shareable but got {type(reply)}")
+            return ReturnCode.EXECUTION_EXCEPTION, None
+
+        rc = reply.get_return_code()
+        if rc != ReturnCode.OK:
+            self.log_error(fl_ctx, f"retrieval request rejected by site {from_site}: {rc}")
+            return rc, None
+
+        # wait for result until either the result is received or progress timed out
+        rc = ReturnCode.OK
+        abort_signal = fl_ctx.get_run_abort_signal()
+        start_time = time.time()
+        while True:
+            # wait a short time so that we can check other conditions
+            if not waiter.wait(_SHORT_WAIT):
+                # see whether we have any progress
+                if time.time() - start_time > timeout:
+                    # no progress for too long
+                    self.log_error(fl_ctx, f"streamn data not completed in {timeout} seconds")
+                    rc = ReturnCode.TIMEOUT
+                    break
+
+                if abort_signal and abort_signal.triggered:
+                    rc = ReturnCode.TASK_ABORTED
+                    break
+            else:
+                # result available!
+                break
+
+        self.tx_table.pop(tx_id, None)
```

Comment:
In case of an exception, there will be a leak. Should do this in final block.

Reply:
Good catch. Fixed.